### DISCUSSION TITLE
Interrupting and resuming execution of tree.

### DISCUSSION THREAD

**Discussion by czajah** [BehaviorTree/BehaviorTree.CPP#635] [Q&A] [answered]
Hi everyone,

I'm working with a behavior tree where an action issues a "GoToPosition" command once, and then continually returns a 'running' status until the agent either reaches the desired position or a timeout occurs.

I'm trying to figure out the best way to implement an interrupt-resume logic. Let's say there's an external failsafe mechanism that can override the last command from the behavior tree in case of danger. Once the danger is avoided, how do I inform the behavior tree that there was an interruption and that the last command needs to be reissued?

One thought is to have the action fail and restart tree from the beginning. But I'm wondering if there's a way to reset the state of the node back to 'idle' and then resume execution from there?

Essentially, I'm looking for a way to smoothly transition between the running tree and failsafe/manual operation states. Any insights would be appreciated!

**Comment by facontidavide** [comment, marked as answer]
Good news: there is a way.
Bad news: YOU are responsible for the implementation.

You can implement this with your overloaded  `halt()` method.

You can save your state/mission when the node is halted and the next time the node is ticked, you resume operations.

**Reply by czajah** [reply]
I've implemented it using events for state transitions. When action is running and there is an event for transition from 'mission' to 'failsafe' im setting flag 'interrupted' and then on transition back from 'failsafe' to 'mission' if 'interrupted' was set then im setting flag 'reissueCommand' and inside tick() I have if(status()==BT::NodeStatus::IDLE || reissueCommand){...}

**Reply by QIAO-666** [reply]
After this reply, I have another question. I can overload halt() in terms of AsyncActionNode, but executing the tree again will tick all nodes after halt() the BT tree. That is not the initial purpose. I think resuming will return to the old status when some nodes had return success and some nodes had return failture，but not all nodes are idle. Is my idea wrong?
